river/ensemble/adaptive_random_forest.py

```diff
@@ -328,7 +333,7 @@ def _new_learning_node(self, initial_stats=None, parent=None):  # noqa
                 initial_stats,
                 depth,
                 self.attr_obs,
-                self.attr_obs_params,
+                attr_obs_params,
```

Comment:
Just an opinion: I prefer to directly pass the attribute observer as a parameter. I don't really like passing a `str` that selects the attribute observer and then passing parameters for said observer on the side. If you look at `LogisticRegression`, the `optimizer` parameter is a `optim.Optimizer`, it's not a `str`. I think it's important to have this discussion, even if we disagree. It's not the end of the world if different parts of the code base take different approaches, but we should at least discuss it.

A nice benefit of directly passing a class instead of a string is that you don't have to document the parameters of the attribute observer in each tree. This is particularly relevant in a "big" code base like ours. The user just knows that they should look at the attribute observer module and take a look at the parameters that are there.

I'd be really interested to know what you think about this. Please tell me if I'm not clear! :)

Reply:
I wish I had thought about that when we first made the attribute observer selection dynamic. It totally makes sense and facilitates the usage of the trees. By using strings and dictionaries we add a lot of clutter to the documentation, and even so, what each attribute observer exactly does is not clear. I would happily change that. What do you think, @jacobmontiel?

But here is where I need some help to plan:

- At this moment, the inner structures of the trees are "private": `_submodules`. How to proceed? Should we make everything public, like splits, nodes, and so on? Only the attribute observers? I am worried about suddenly adding too much information to the user process (documentation-wise).
- Secondly, and maybe more critical, E-BST and TE-BST would easily work under this new approach. It would be just a matter of using a clone of sorts or deep copies to populate the tree leaves. QO, on the other hand, has different parameters passed in runtime. I am not sure how the tree could modify the radius passed to QO in an online fashion. Here' some ideas:
    * Use `set_params`?
    * *(I don't like this one very much)* Make the radius static: we could require the user to apply a normalization step (StandardScaler or MinMaxScaler) when using QO. I know, I know, the user already does that because model tree leaves work better when all features are scaled. But, if the mean predictor is used instead, it would be nice to have the freedom to avoid an unnecessary preprocessing step and infer the adequate quantization radius on the go.

What do you think guys? @online-ml/devs, any feedback is welcomed!

Reply:
> At this moment, the inner structures of the trees are "private": _submodules. How to proceed? Should we make everything public, like splits, nodes, and so on? Only the attribute observers? I am worried about suddenly adding too much information to the user process (documentation-wise).

I understand. I would only make public the stuff that the user will specify. I'm not 100% aware of all the parameters you're exposing, but having a `split` module where you put all the attribute observers (I call them splitters) seems like a good idea.

> Secondly, and maybe more critical, E-BST and TE-BST would easily work under this new approach. It would be just a matter of using a clone of sorts or deep copies to populate the tree leaves. QO, on the other hand, has different parameters passed in runtime. I am not sure how the tree could modify the radius passed to QO in an online fashion. Here' some ideas: 

I guess your ideas make sense. You could inherit from `base.Base` and that will simplify your life.

Reply:
Inheriting from `base.Base` might do the trick. Should I address such a change in our codebase in this PR? I could also proceed with these changes in another PR to limit the scope. There we can discuss naming convention and documentation. 

What do you think @MaxHalford and @jacobmontiel?

Reply:
I would say do what feels more comfortable for you :)

Reply:
Cool! Let's wait for Jacob's feedback :D

Reply:
Sorry for the delay. I also prefer passing an object rather than a string. In the end, this is an advanced option and makes sense to give more freedom/responsibility to the user.

We are also using this approach is in the `ARF` and `SRP` where the user can pass a drift detector ;-)

> You could inherit from `base.Base` and that will simplify your life.

+1

> I would only make public the stuff that the user will specify.

+1 We can always expose other things down the road as long as it is necessary.

> Should I address such a change in our codebase in this PR? I could also proceed with these changes in another PR to limit the scope.

I recommend using a different PR because

1. The change may impact other methods outside of the current scope of this PR. Historically, we have seen this happening on similar refactorings.
2. There are non-trivial issues to address like the passing of parameters at runtime for the QO.

Let me know if this is useful.